spatial/r3/box.go

```diff
@@ -0,0 +1,79 @@
+// Copyright ©2022 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package r3
+
+// Box is a 3D bounding box.
+type Box struct {
+	Min, Max Vec
+}
+
+// CenteredBox creates a Box with a given center and size.
+func CenteredBox(center, size Vec) Box {
+	half := Scale(0.5, size)
+	return Box{Min: Sub(center, half), Max: Add(center, half)}
+}
```

Comment:
How often is this needed? It seems to me that it wouldn't be often and that it's simple enough to leave to users. (Though see comment on constructor).

Reply:
I'd be ok with removing this. That said this form seems to be the most common form of constructing a box from scratch from what I've been doing.
What you are suggesting below is having a `NewBox(a,b Vec)` that ensures valid box creation, right?

Reply:
> What you are suggesting below is having a `NewBox(a,b Vec)` that ensures valid box creation, right?

Yes, so that `b.Min` is actually `elemMin(a, b)` and `b.Max` is `elemMax(a, b)`.

Reply:
Alright, I've added Empty method along with valid box creation algorithm for `CenteredBox`, `Union` and `Scale`.

I still feel like `CenteredBox` is the most reasonable way of initializing a Box where complexity is reasonably justfied.

I'll also comment on `NewBox`: A `NewBox` constructor which does the same thing as `Rect` is not intuitive to me at all. "NewBox creates a box that contains all components of the two vectors"? In every case where I've created a box I've had precise information on how to create `Min` and `Max` from scratch. I feel like going the `Rect` way lends itself to messy Box creation where the user hopes the resulting Box contains the desired bounds.

Reply:
The constructor defines the box by the position of its side rather than the position of two of its corners. This is the intuition. An alternative way to think about it is as the union of two zero-sized boxes at the defined min and max (which many not have been min and max).

Reply:
I do not see the difference between defining a box by its side over its corner. The corners lie on the sides after all? 

Let me restate my fear: If we allow constructing valid boxes with two arbitrary vectors then a lot of readability can be lost when it is misused. When I say misuse I mean users can find clever ways of constructing boxes such that in the example`NewBox(a,b)` `a.X < b.X` while `a.Y>b.Y`, resulting in confusion over which vector defines which side/corner (at least for me). When constructing boxes with `Box{Min:a, Max:b}` the intention is clear and the code more structured.

Reply:
> I do not see the difference between defining a box by its side over its corner. The corners lie on the sides after all?

Taking r2 to reduce complsxit of example, Box{Min:Vec{X:1,Y:1}, Max:Vec{2,2}} does not work the same as Box{Min:Vec{X:1,Y:2}, Max:Vec{2,1}} (this would be an invalid box under the implementations that I was looking at yesterday). In pseudo code Box for X=1..2 Y=1..2 is guaranteed to mean what the user intends.

> If we allow constructing valid boxes with two arbitrary vectors

They are not vectors. Take a look at [`image.Rect`](https://pkg.go.dev/image#Rect); it accepts integer arguments. `NewBox` would be `NewBox(x0, y0, x1, y1 float64) Box`. These parameters merely say make a box with vertical sides at x0 and x1 and horizontal sides at y0 and y1.

Reply:
OK, if we are talking about a function signature like `NewBox(x0, y0, z0, x1, y1, z1 float64)` then I can agree this is OK since now it is clear the box is being created with sides rather than points (as implied by using `Vec` type). I was initially against this signature because it had lots of arguments- however, looking through my use of `Box` it's clear this function would replace **many** `Box{Min...Max...}` initializations.
Little question on the signature: which do we prefer?
```go
NewBox(x0, y0, z0, x1, y1, z1 float64)
NewBox(x0, x1, y0, y1, z0, z1 float64)
```
I am partial to the second one since related numbers are side-by-side.

Reply:
While I prefer the second, I'm going to argue for the first on the basis of similarity with `image.Rect`.